core/transaction-pool/graph/src/validated_pool.rs

```diff
@@ -182,18 +187,100 @@ impl<B: ChainApi> ValidatedPool<B> {
 		}
 	}
 
+	/// Resubmits transactions in the pool.
+	///
+	/// Removes and then submits passed transactions and all dependent transactions.
+	/// Transactions that are missing from the pool are not submitted.
+	pub fn resubmit(&self, mut transactions: HashMap<ExHash<B>, ValidatedTransactionFor<B>>) {
+		#[derive(Debug, Clone, Copy, PartialEq)]
+		enum Status { Future, Ready, Failed, Dropped };
+
+		let (initial_statuses, final_statuses) = {
+			let mut pool = self.pool.write();
+
+			// remove all passed transactions from the ready/future queues
+			// (this may remove additional transactions as well)
+			let mut initial_statuses = HashMap::new();
+			let mut txs_to_resubmit = Vec::with_capacity(transactions.len());
+			while !transactions.is_empty() {
+				let hash = transactions.keys().next().cloned().expect("transactions is not empty; qed");
+
+				let removed = pool.remove_from_ready(&hash);
+				for removed_tx in removed {
+					let removed_hash = removed_tx.hash.clone();
+					let tx_to_resubmit = if let Some(updated_tx) = transactions.remove(&removed_hash) {
```

Comment:
I'm not sure I fully understand the logic here. So we can have a different transaction under `removed_hash` it seems, but yet we never really store it's hash anywhere and don't notify listeners about `hash(updated_tx)` only about `hash(removed_tx)`.

This may lead to a situation where `updated_tx` get's accepted, and we actually send an info that `hash(removed_tx)` is now ready.

Reply:
It maybe my misuderstanding as well. I don't think that we can have different transaction under `remove_hash`. And I don't clearly understand what do you mean by "we never really store it's hash anywhere and don't notify ...".. Let me describe this while with example and check that we both on the same page. Let's say at some moment we had a pool with:

HashOf(Tx1) => { data=Tx1, requires=[], provides=[1] }
HashOf(Tx2) => { data=Tx2, requires=[1], provides=[2] }
HashOf(Tx3) => { data=Tx3, requires=[], provides=[3] }

At this moment light client decides to revalidate its transactions. It sends revalidation requests for [Tx1, Tx2, Tx3] to remote nodes. Let's say while we wait for the response, another transactions are inserted to the pool:

HashOf(Tx4) => { data=Tx4, requires=[3], provides=[] }
HashOf(Tx5) => { data=Tx5, requires=[], provides=[] }

Let's say we have received responses (aka updated validation results):

HashOf(Tx1) => { data=Tx1, requires=[3], provides=[] }
HashOf(Tx2) => is now considered invalid
HashOf(Tx3) => { data=Tx3, requires=[], provides=[3] }

Then the `resubmit()` is called with these results. The `while !transactions.is_empty() {}` we're talking about must remove transactions Tx1, Tx2, Tx3 along with all transactions that depend on these tx. In our case they're: Tx4. While removing, this `while` builds a map of `TxHash => Either<UpdatedValidationResult, PreviousValidationResult>`. In our case Tx1, Tx2, Tx3 will have `UpdatedValidationResult` and Tx4 will have `PreviousValidationResult`. So the resulting map is:

HashOf(Tx1) => { data=Tx1, requires=[3], provides=[] }
HashOf(Tx2) => is now considered invalid
HashOf(Tx3) => { data=Tx3, requires=[], provides=[3] }
HashOf(Tx4) => { data=Tx4, requires=[3], provides=[] }

(mind that Tx5 is not affected by resubmit at all and stays in pool). So we're always referring to the same transactions with their hashes. Could you, please, check my logic here - maybe I'm missing something you're talking about?

Reply:
(While responding to this, I've found that the order of insertion in `resubmit()` makes sense if we reject 'future' transactions. I.e. if we'll try to insert Tx1 before Tx3, then Tx1 will be rejected. So we either need to: (1) insert in the correct order (2) clear 'future' queue at the end of `resubmit()` call (with proper notifications, of course) or (3) support future queue on light client. I'd still prefer to avoid future transactions here => will try to go with (2) here)

Reply:
Thanks for explanations! Yes, that totally makes sense, for some reason I was confused with what exactly we store in the `HashMap<Hash, ValidatedTransaction>`. Maybe renaming `transactions` to `updated_transactions` and adding a short comment that we store re-validated transactions there would make it clearer.

Re futures: indeed `(2)` seems like a better way to go, so you would temporarily need to allow `futures` queue for this specific calls and then just clear it after.